Allow sending headers to the elasticsearch client

alexandrustaetu opened the pull request:
## Pull Request Checklist

** #103  **

#### General

- [x] Update Changelog following the conventions laid out [here](https://github.com/sensu-plugins/community/blob/master/HOW_WE_CHANGELOG.md)

- [ ] Update README with any necessary configuration snippets

- [ ] Binstubs are created if needed

- [ ] RuboCop passes

- [ ] Existing tests pass

#### New Plugins

- [ ] Tests

- [ ] Add the plugin to the README

- [ ] Does it have a complete header as outlined [here](http://sensu-plugins.io/docs/developer_guidelines.html#coding-style)

#### Purpose

#### Known Compatibility Issues

majormoses commented:
Thanks for your contribution to Sensu plugins! Without people like you submitting PRs we couldn't run the project. I will review it shortly.

majormoses commented:
can we use headers and allow it to be a comma separated list of headers? I am sure at some point we are gonna need more than one header. Also can you remove the extra newline to appease the cops.

majormoses commented on bin/check-es-query-ratio.rb:
```diff
@@ -179,6 +179,11 @@ class ESQueryRatio < Sensu::Plugin::Check::CLI
          short: '-u USER',
          long: '--user USER'
 
+  option :header,
```
headers

majormoses commented on bin/check-es-query-ratio.rb:
```diff
@@ -179,6 +179,11 @@ class ESQueryRatio < Sensu::Plugin::Check::CLI
          long: '--user USER'
 
+  option :header,
+         description: 'Headers to pass to the elasticsearch http client.',
```
A comma separated list of headers to pass to elasticsearch http client

majormoses commented on bin/check-es-query-ratio.rb:
```diff
@@ -179,6 +179,11 @@ class ESQueryRatio < Sensu::Plugin::Check::CLI
 
+  option :header,
+         description: 'Headers to pass to the elasticsearch http client.',
+         short: '-H header',
```
headers

majormoses commented on bin/check-es-query-ratio.rb:
```diff
@@ -179,6 +179,11 @@ class ESQueryRatio < Sensu::Plugin::Check::CLI
+  option :header,
+         description: 'Headers to pass to the elasticsearch http client.',
+         short: '-H header',
+         long: '--header header'
```
headers

majormoses commented on lib/sensu-plugins-elasticsearch/elasticsearch-common.rb:
```diff
@@ -43,6 +43,22 @@ def client
-    @client ||= Elasticsearch::Client.new(transport_class: transport_class, hosts: [host], region: config[:region])
+    transport_options = {}
+
+    if config[:header]
```
headers

majormoses commented on lib/sensu-plugins-elasticsearch/elasticsearch-common.rb:
```diff
@@ -43,6 +43,22 @@ def client
+
+      headers = {}
+
+      config[:header].split(',').each do |header|
```
headers

majormoses commented on bin/check-es-query-exists.rb:
```diff
@@ -157,6 +157,11 @@ class ESQueryExists < Sensu::Plugin::Check::CLI
          long: '--user USER'
 
+  option :header,
+         description: 'Headers to pass to the elasticsearch http client.',
```
A comma separated list of headers to pass to elasticsearch http client

majormoses commented on bin/check-es-query-exists.rb:
```diff
@@ -157,6 +157,11 @@ class ESQueryExists < Sensu::Plugin::Check::CLI
 
+  option :header,
+         description: 'Headers to pass to the elasticsearch http client.',
+         short: '-H header',
```
headers

majormoses commented on bin/check-es-query-exists.rb:
```diff
@@ -157,6 +157,11 @@ class ESQueryExists < Sensu::Plugin::Check::CLI
+  option :header,
+         description: 'Headers to pass to the elasticsearch http client.',
+         short: '-H header',
+         long: '--header header'
```
headers

majormoses commented on bin/check-es-query-count.rb:
```diff
@@ -171,6 +171,11 @@ class ESQueryCount < Sensu::Plugin::Check::CLI
          long: '--user USER'
 
+  option :header,
+         description: 'Headers to pass to the elasticsearch http client.',
```
A comma separated list of headers to pass to elasticsearch http client

majormoses commented on bin/check-es-query-count.rb:
```diff
@@ -171,6 +171,11 @@ class ESQueryCount < Sensu::Plugin::Check::CLI
 
+  option :header,
+         description: 'Headers to pass to the elasticsearch http client.',
+         short: '-H header',
```
headers

majormoses commented on bin/check-es-query-count.rb:
```diff
@@ -171,6 +171,11 @@ class ESQueryCount < Sensu::Plugin::Check::CLI
+  option :header,
+         description: 'Headers to pass to the elasticsearch http client.',
+         short: '-H header',
+         long: '--header header'
```
headers

majormoses commented on bin/check-es-query-average.rb:
```diff
@@ -183,6 +183,11 @@ class ESQueryAverage < Sensu::Plugin::Check::CLI
          long: '--user USER'
 
+  option :header,
+         description: 'Headers to pass to the elasticsearch http client.',
```
A comma separated list of headers to pass to elasticsearch http client

majormoses commented on bin/check-es-query-average.rb:
```diff
@@ -183,6 +183,11 @@ class ESQueryAverage < Sensu::Plugin::Check::CLI
          short: '-u USER',
          long: '--user USER'
 
+  option :header,
```
headers

majormoses commented on bin/check-es-query-average.rb:
```diff
@@ -183,6 +183,11 @@ class ESQueryAverage < Sensu::Plugin::Check::CLI
 
+  option :header,
+         description: 'Headers to pass to the elasticsearch http client.',
+         short: '-H header',
```
headers

majormoses commented on bin/check-es-query-average.rb:
```diff
@@ -183,6 +183,11 @@ class ESQueryAverage < Sensu::Plugin::Check::CLI
+  option :header,
+         description: 'Headers to pass to the elasticsearch http client.',
+         short: '-H header',
+         long: '--header header'
```
headers

majormoses commented on bin/check-es-query-count.rb:
```diff
@@ -171,6 +171,11 @@ class ESQueryCount < Sensu::Plugin::Check::CLI
          short: '-u USER',
          long: '--user USER'
 
+  option :header,
```
headers

majormoses commented on bin/check-es-query-exists.rb:
```diff
@@ -157,6 +157,11 @@ class ESQueryExists < Sensu::Plugin::Check::CLI
          short: '-u USER',
          long: '--user USER'
 
+  option :header,
```
headers

alexandrustaetu commented:
Hello,

Did new line trimming, added default Content-Type header for the changed bins and changed the parameter name as requested. 

I'll also try to test today if the default content-type header set to application-json breaks interaction with ES 2,3,4,5.

Regarding the testing artifact, how would it be appropriate to proceed with? I saw that https://github.com/sensu-plugins/sensu-plugins-dns repo has a .kitchen.yml file and a test dir with a set of tests used for automated testing. Should the same file structure / testing be created / added to this repo too?

Thanks,
Alex.

alexandrustaetu commented:
As a work in progress for the automated tests, I followed the same procedure used on sensu-plugins-dns and added a .kitchen.yaml file that runs 2 tests.
Considering the volume of changes this brings, should a different pull request be created for this?

Changes related to automated tests can be found [here](https://github.com/alexandrustaetu/sensu-plugins-elasticsearch/compare/master...alexandrustaetu:automated_testing_via_kitchen?expand=1):

majormoses commented:
I am fine doing either adding it here or doing it as a separate PR. I really appreciate your willingness to write tests! If you need this out quickly just provide a gist showing it works and we can work on the automated tests working otherwise I would love to use this opportunity as you are willing to get some tests going. I find people are more willing to write tests for plugins when there is an example already in the repo. If I had enough time I would run through all the plugins and add them myself but my gree time is limited. I have run through a handful to at least give examples to allow the community to help crowdsource it.

alexandrustaetu commented:
Hello @majormoses ,

Added testing to the current pull with a description in readme on how to run for future contributors.

As requested, I also added a gist with output for successfully testing on this repo.

Let me know if anything is in place for the proposed changes.

Thanks a lot,
Alex.

majormoses commented:
I think I added the missing bits to make it work in travis I tested locally and it looks fantastic thank you for your efforts! After CI passes I will merge and release it.

majormoses commented:
Looks like they all passed except ruby 2.0 as there were not a suite for it. Will fixup.

majormoses commented:
released: https://rubygems.org/gems/sensu-plugins-elasticsearch/versions/1.11.0

alexandrustaetu commented:
Thanks, @majormoses , this pull request was resolved really fast. Seeing how easy it is to contribute, I'll try to create a set of pull requests for other sensu plugins to add integration testing for the upcoming days.

Thanks for your time!

majormoses commented:
Awesome I really appreciate that and am looking forward to reviewing them!
